x/mongo/driver/topology/pool.go

```diff
@@ -358,163 +299,103 @@ func (p *pool) get(ctx context.Context) (*connection, error) {
-			// more resources than we previously thought, so we try to get a resource again.
-			made := p.conns.incrementTotal()
-			if !made {
-				continue
-			}
-			c, reason, err := p.makeNewConnection()
-
-			if err != nil {
-				if p.monitor != nil {
-					// We only publish a GetFailed event because makeNewConnection has already published
-					// ConnectionClosed if needed.
-					p.monitor.Event(&event.PoolEvent{
-						Type:    event.GetFailed,
-						Address: p.address.String(),
-						Reason:  reason,
-					})
-				}
-				p.conns.decrementTotal()
-				p.sem.Release(1)
-				return nil, err
-			}
-
-			c.connect(ctx)
-			// wait for conn to be connected
-			err = c.wait()
-			if err != nil {
-				// Call removeConnection to remove the connection reference and fire a ConnectionClosedEvent.
-				_ = p.removeConnection(c, event.ReasonConnectionErrored)
-				p.conns.decrementTotal()
-				p.sem.Release(1)
-
-				if p.monitor != nil {
-					p.monitor.Event(&event.PoolEvent{
-						Type:    event.GetFailed,
-						Address: p.address.String(),
-						Reason:  event.ReasonConnectionErrored,
-					})
-				}
-				return nil, err
-			}
+			err = w.err
+			return
+		}
 
-			if p.monitor != nil {
-				p.monitor.Event(&event.PoolEvent{
-					Type:         event.GetSucceeded,
-					Address:      p.address.String(),
-					ConnectionID: c.poolID,
-				})
-			}
-			return c, nil
+		if p.monitor != nil {
+			p.monitor.Event(&event.PoolEvent{
+				Type:         event.GetSucceeded,
+				Address:      p.address.String(),
+				ConnectionID: w.conn.poolID,
+			})
+		}
+		conn = w.conn
+		return
+	case <-ctx.Done():
+		if p.monitor != nil {
+			p.monitor.Event(&event.PoolEvent{
+				Type:    event.GetFailed,
+				Address: p.address.String(),
+				Reason:  event.ReasonTimedOut,
+			})
+		}
+		err = WaitQueueTimeoutError{
+			Wrapped:                      ctx.Err(),
+			PinnedCursorConnections:      atomic.LoadUint64(&p.pinnedCursorConnections),
+			PinnedTransactionConnections: atomic.LoadUint64(&p.pinnedTransactionConnections),
+			maxPoolSize:                  p.maxSize,
+			totalConnectionCount:         p.totalConnectionCount(),
 		}
+		return
 	}
 }
 
-// closeConnection closes a connection, not the pool itself. This method will actually closeConnection the connection,
-// making it unusable, to instead return the connection to the pool, use put.
-func (p *pool) closeConnection(c *connection) error {
-	if c.pool != p {
+// closeConnection closes a connection.
+func (p *pool) closeConnection(conn *connection) error {
+	if conn.pool != p {
 		return ErrWrongPool
 	}
 
-	if atomic.LoadInt64(&c.connected) == connected {
-		c.closeConnectContext()
-		_ = c.wait() // Make sure that the connection has finished connecting
-	}
-
-	if !atomic.CompareAndSwapInt64(&c.connected, connected, disconnected) {
-		return nil // We're closing an already closed connection
+	if atomic.LoadInt64(&conn.connected) == connected {
```

Comment:
You're correct that `connecting` is not used as a `connection` state. The "connected state" constants that `connection`, `pool`, `Server`, and `Topology` use are the same, but the state transitions are confusingly different. I've created [GODRIVER-2181](https://jira.mongodb.org/browse/GODRIVER-2181) as a suggested improvement to standardize those state transitions to avoid confusion (and fix a few bugs I found while investigating it).